Source/Scene/ModelExperimental/CustomShaderStage.js

```diff
@@ -0,0 +1,519 @@
+  }
+
+  var alphaOptions = renderResources.alphaOptions;
+  if (customShader.isTranslucent) {
+    alphaOptions.pass = Pass.TRANSLUCENT;
+    alphaOptions.alphaMode = AlphaMode.BLEND;
+  } else {
+    // Use the default pass (either OPAQUE or 3D_TILES), regardless of whether
+    // the material pipeline stage used translucent. The default is configured
+    // in AlphaPipelineStage
+    alphaOptions.pass = undefined;
+    alphaOptions.alphaMode = AlphaMode.OPAQUE;
+  }
+
+  renderResources.uniformMap = combine(
+    renderResources.uniformMap,
+    customShader.uniformMap
+  );
+};
+
+function getAttributeNames(attributes) {
+  var names = {};
+  for (var i = 0; i < attributes.length; i++) {
+    var attribute = attributes[i];
+    var semantic = attribute.semantic;
+    var setIndex = attribute.setIndex;
+
+    var variableName;
+    if (defined(semantic)) {
+      variableName = VertexAttributeSemantic.getVariableName(
+        semantic,
+        setIndex
+      );
+    } else {
+      // Handle user defined vertex attributes. They must begin with an underscore
+      // For example, "_TEMPERATURE" will be converted to "a_temperature".
+      variableName = attribute.name.substring(1).toLowerCase();
+    }
+
+    names[variableName] = attribute;
+  }
+  return names;
+}
+
+function generateStructField(glslType, fieldName) {
+  return "    " + glslType + " " + fieldName + ";";
+}
+
+function generateAttributeField(name, attribute) {
+  var attributeType = attribute.type;
+  var glslType = AttributeType.getGlslType(attributeType);
+
+  // Fields for the Attribute struct. for example:
+  // "    vec3 position;"
+  return generateStructField(glslType, name);
+}
+
+// GLSL types of standard attribute types when uniquely defined
+var attributeTypeLUT = {
+  position: "vec3",
+  normal: "vec3",
+  tangent: "vec4",
+  texCoord: "vec2",
+  joints: "ivec4",
+  weights: "vec4",
+};
+
+// Corresponding attribute values
+var attributeDefaultValueLUT = {
+  position: "vec3(0.0)",
+  normal: "vec3(0.0, 0.0, 1.0)",
+  tangent: "vec4(1.0, 0.0, 0.0, 1.0)",
+  texCoord: "vec2(0.0)",
+  joints: "ivec4(0)",
+  weights: "vec4(0.0)",
+};
+
+function inferAttributeDefaults(attributeName) {
+  // remove trailing set indices. E.g. "texCoord_0" -> "texCoord"
+  var trimmed = attributeName.replace(/_[0-9]+$/, "");
+  var glslType = attributeTypeLUT[trimmed];
+  var value = attributeDefaultValueLUT[trimmed];
+
+  // Return undefined for other cases that cannot be easily inferred:
+  // - COLOR_x is either a vec3 or vec4
+  // - _CUSTOM_ATTRIBUTE has an unknown type.
+  if (!defined(glslType)) {
+    return undefined;
+  }
+
+  return {
+    attributeField: generateStructField(glslType, attributeName),
+    value: value,
+  };
+}
+
+function generateVertexShaderLines(customShader, namedAttributes) {
+  var categories = partitionAttributes(
+    namedAttributes,
+    customShader._usedVariablesVertex.attributeSet
```

Comment:
Should `_usedVariablesVertex` be a non-underscore property?

Reply:
hm yeah I can change it to be a regular property that's marked `@private`